projects/bootloader/test/test_query.c

```diff
@@ -0,0 +1,280 @@
+  // "project_info" : ["abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz12345678910"]
+  // "git_version" : ["abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz12345678910"]
+  size_t len = 262;
+  // exact query, so this is also the expected response
+  uint8_t data[262] = {
+    0x08, 0x02, 0x12, 0x3F, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C,
+    0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x61, 0x62,
+    0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72,
+    0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38,
+    0x39, 0x31, 0x30, 0x1A, 0x3F, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B,
+    0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x61,
+    0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71,
+    0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+    0x38, 0x39, 0x31, 0x30, 0x22, 0x3F, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A,
+    0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A,
+    0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70,
+    0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36,
+    0x37, 0x38, 0x39, 0x31, 0x30, 0x2A, 0x3F, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69,
+    0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79,
+    0x7A, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F,
+    0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7A, 0x31, 0x32, 0x33, 0x34, 0x35,
+    0x36, 0x37, 0x38, 0x39, 0x31, 0x30,
+  };
+  // response
+  size_t expected_len = len;
+  uint8_t *expected_response = data;
+}
+
+void test_repeated_field(void) {
+  BootloaderConfig config = {
+    .controller_board_id = s_board_id,
+    .controller_board_name = "hello",
+    .project_present = true,
+    .project_name = "centre console",
+    .project_info = "testing",
+    .git_version = "random info",
+  };
+  query_init(&config);
+
+  // "id" : [ 1, 2, 3 ]
+  // "name" : [ "hello", "world" ]
+  // "current_project" : [ "pedal board", "centre console" ]
+  // "project_info" : ["testing"]
+  // "git_version" : []
+  size_t query_len = 58;
+  uint8_t query[58] = {
+    0x08, 0x01, 0x08, 0x02, 0x08, 0x03, 0x12, 0x05, 0x68, 0x65, 0x6C, 0x6C, 0x6F, 0x12, 0x05,
+    0x77, 0x6F, 0x72, 0x6C, 0x64, 0x1A, 0x0B, 0x70, 0x65, 0x64, 0x61, 0x6C, 0x20, 0x62, 0x6F,
+    0x61, 0x72, 0x64, 0x1A, 0x0E, 0x63, 0x65, 0x6E, 0x74, 0x72, 0x65, 0x20, 0x63, 0x6F, 0x6E,
+    0x73, 0x6F, 0x6C, 0x65, 0x22, 0x07, 0x74, 0x65, 0x73, 0x74, 0x69, 0x6E, 0x67,
+  };
```

Comment:
How did you generate these hexdumps? Could you generate them with nanopb inside the test instead of declaring these raw hexdumps? Currently it's impossible to edit by hand or verify that the data is what the comments say it is. (Also, pro tip: you can just declare `uint8_t query[] = {...}` and C will fill in the length for you.)